### ISSUE TITLE
Gradient Clipping

### ISSUE THREAD

**Issue by rizar** [bartvm/blocks#185]
When designing the `StepRule`'s interface I made a mistake: I forgot that for gradient clipping one needs access to gradients w.r. all parameters at the same time, not only to each one separately. Thus `StepRule` should be redesigned a bit to make gradient clipping possible.

**Comment by bartvm** [comment]
+1 Because I ran into this same issue yesterday when implementing a line search for something, it's nicer if the step rule accepts all the parameters at once. In general, we shouldn't gloss over Pylearn2's way of doing things here, the learning rules/algorithms there have never been a major point of concern as far as I know, and they do that [as well](https://github.com/lisa-lab/pylearn2/blob/master/pylearn2/training_algorithms/learning_rule.py#L129-L164).

**Comment by rizar** [comment]
I am fine with PyLearn2'ish way (as of now), except that the learning rate belongs to the step rule in our case, not to the algorithm

**Comment by rizar** [comment]
A quick note: we should be ready that step rules should be chainable. E.g., in Groundhog algorithms the gradient was first cut off and then accumulated.
